### PR TITLE
adding fees to ledger spec

### LINKED ISSUE #50: models fee in the leger
We should model the transaction fees in the ledger spec.

### PR DESCRIPTION
**Pull request by JaredCorduan**
closes #50

### CODE REVIEW

File: specs/ledger/latex/ledger-spec.tex [hunk 1]
```diff
@@ -300,8 +306,8 @@ \section{UTxO}
 \begin{figure}
   \begin{equation}\label{eq:utxo-inductive}
     \inference[UTxO-inductive]
-    { \txins{tx} \subseteq \dom \var{utxo}
-      & \balance{(\txouts{tx})} \leq \balance{(\txins{tx} \restrictdom \var{utxo})}
+    { \txins{tx} \subseteq \dom \var{utxo} & \minfee{tx} \leq \txfee{tx}\\
```
Comment (dnadales) [inline specs/ledger/latex/ledger-spec.tex]: According [this message](https://input-output-rnd.slack.com/archives/CCZSDE005/p1540233076000100), it seems `minfee` depends on parameters outside the transaction `tx`, called `a` and `b`. So it seems we'd need to add a transition environment, that contains `a` and `b` and model this is `minfee a b tx` (I need to look what `a` and `b` are).

File: specs/ledger/latex/ledger-spec.tex [hunk 2]
```diff
@@ -325,6 +331,15 @@ \section{UTxO}
   $\var{tx}$.
 \end{itemize}
 
+\begin{note}
```
Comment (dnadales) [inline specs/ledger/latex/ledger-spec.tex:339]: That's indeed a good point! So what you're saying is that if we'd take coin to be `Z` then we could spend more than what we have by having negative outputs, right? Maybe we should either put constraints on the `Coin` type (e.g. commutative monoid), or specify in the rule that all values should be `>= 0` (but the `Coin` is not abstract anymore). Should we try to fix this in this pull request?

**Review by dnadales — COMMENTED** [review]
Thanks a lot for this!

I think we should address:

0. Modelling the parameters on which the `minfee` function depends.
1. Restrict the possible instances of `Coin`.

File: specs/ledger/latex/ledger-spec.tex (hunk 2 above)
Comment (nc6) [reply specs/ledger/latex/ledger-spec.tex:339]: At the moment, `Coin` is given no specification (which is a problem since we try to sum over it in the `balance` operation!) I feel we should constrain the coin type.

**Review by nc6 — APPROVED** [review]

**Comment by nc6** [comment]
I had no comments aside from Damian's, so giving approval. I think we should fix the `minfee` params, though I'd be happy to address the `Coin` issue in a separate PR.

File: specs/ledger/latex/ledger-spec.tex [hunk 3]
```diff
@@ -242,7 +244,11 @@ \section{UTxO}
       \txid{} & \Tx \to \TxId & \text{compute transaction id}\\
       %
       \fun{txbody} & \Tx \to \powerset{\TxIn} \times (\Ix \mapsto \TxOut)
-                                  & \text{transaction body}
+                                  & \text{transaction body}\\
+      %
+      \fun{txfee} & \Tx \to \Coin & \text{transaction fee}\\
+      %
+      \fun{minfee} & \Tx \to \Coin & \text{minimum fee}
```
Comment (mgudemann) [inline specs/ledger/latex/ledger-spec.tex]: according to a message from Duncan, close to the Damien's link below, we might make this a function from the size of `Tx` to `Coin`, as it depends solely on the size and not the content of the transaction.

Comment (mgudemann) [reply specs/ledger/latex/ledger-spec.tex]: that is having type `N -> Coin`

**Comment by dnadales** [comment]
> I had no comments aside from Damian's, so giving approval. I think we should fix the `minfee` params, though I'd be happy to address the `Coin` issue in a separate PR.

Sounds good @nc6! I've created an [issue](https://github.com/input-output-hk/cardano-chain/issues/54) about this.

File: specs/ledger/latex/ledger-spec.tex (hunk 2 above)
Comment (JaredCorduan) [reply specs/ledger/latex/ledger-spec.tex:339]: yea, that is my preference as well, to make it essentially N.  Commutative monoid is unfortunately not enough (Z meets that as well).  I'm happy to change this here. I just put it as a note since I wasn't sure if it required a larger conversation.  Eventually, once we introduce Chimeric types and therefore need subtraction or inverses, the decision will be trickier.

File: specs/ledger/latex/ledger-spec.tex (hunk 1 above)
Comment (JaredCorduan) [reply specs/ledger/latex/ledger-spec.tex]: that's a great point @dnadales , thanks.

File: specs/ledger/latex/ledger-spec.tex (hunk 3 above)
Comment (JaredCorduan) [reply specs/ledger/latex/ledger-spec.tex]: I left it as `Tx` since that we the spec reads the same whether or not the fee is implicit (ie the difference) or explicit.

Comment (JaredCorduan) [reply specs/ledger/latex/ledger-spec.tex]: but perhaps it is time to let go of trying to keep this ledger spec the same as the others yet to come.

**Review by dnadales — APPROVED** [review]

**Comment by JaredCorduan** [comment]
@dnadales - thanks again for catching the missing protocol constants!  I've added them to the model, including the chain extension rule. it's probably worth another good look, it is not as minimal a change as it was before...

File: specs/ledger/latex/blockchain.tex [hunk 4]
```diff
@@ -22,6 +22,14 @@ \subsection{Chain extension}
 outputs in the context of a block are given in
 Sections~\ref{sec:block-delegation} and \ref{sec:block-utxo} respectively.
 
+\begin{note}
+  The protocal constants are currently modeled as a part of the
```
Comment (dnadales) [inline specs/ledger/latex/blockchain.tex]: Protocal sounds like the first calendar humans ever used :P (sorry it is early in the morning)

File: specs/ledger/latex/blockchain.tex [hunk 5]
```diff
@@ -22,6 +22,14 @@ \subsection{Chain extension}
...
+  chain extension environment. Once we adding voting
+  (the mechanism which controls the protocol constants)
+  to the model, it will probably make sense to move them to the
+  chain extension states.
+\end{note}
```
Comment (dnadales) [inline specs/ledger/latex/blockchain.tex:41]: Actually I would say that the chain extension state will include these "constants", but the ledger rules that model voting and updates will have them as well.

File: specs/ledger/latex/ledger-spec.tex [hunk 6]
```diff
@@ -306,10 +310,11 @@ \section{UTxO}
 \begin{figure}
   \begin{equation}\label{eq:utxo-inductive}
     \inference[UTxO-inductive]
-    { \txins{tx} \subseteq \dom \var{utxo} & \minfee{tx} \leq \txfee{tx}\\
-      \balance{(\txouts{tx})}  + \txfee{tx} = \balance{(\txins{tx} \restrictdom \var{utxo})}
+    { \txins{tx} \subseteq \dom \var{utxo} & \minfee{pc}{tx} \leq \txfee{tx}\\
+      \balance{(\txouts{tx})}  + \txfee{tx} =
```
Comment (dnadales) [inline specs/ledger/latex/ledger-spec.tex:314]: So the only `minfee` depends on the protocol (fee) constants, right?

**Review by dnadales — APPROVED** [review]
Looks good! Thanks!

File: specs/ledger/latex/blockchain.tex (hunk 4 above)
Comment (JaredCorduan) [reply specs/ledger/latex/blockchain.tex]: eventually outdated by cro-magnon-cal :calendar:

File: specs/ledger/latex/ledger-spec.tex (hunk 6 above)
Comment (JaredCorduan) [reply specs/ledger/latex/ledger-spec.tex:314]: as of now, yes

File: specs/ledger/latex/blockchain.tex (hunk 5 above)
Comment (JaredCorduan) [reply specs/ledger/latex/blockchain.tex:41]: So would you reword this as

> Once we adding voting (the mechanism which controls the protocol constants) to the model, it will probably make sense to move them to the chain extension states and add them to the rules involving votes

?
